src/display.rs

```diff
@@ -78,46 +77,89 @@ impl Display for InlineTable {
     }
 }
 
-struct TableFormatState<'t> {
-    path: RefCell<Vec<&'t str>>,
-    table: Cell<&'t Table>,
-}
-
-impl<'a> Display for TableFormatState<'a> {
-    fn fmt(&self, f: &mut Formatter) -> Result {
-        let table = self.table.get();
-
-        for kv in table.items.values() {
-            if let Item::Value(ref value) = kv.value {
-                writeln!(f, "{}={}", kv.key, value)?;
+impl Table {
+    fn visit_nested_tables_in_original_order<'t, F>(
+        &'t self,
+        path: &mut Vec<&'t str>,
+        is_array_of_tables: bool,
+        callback: &mut F,
+    ) -> Result
+    where
+        F: FnMut(&Table, &Vec<&'t str>, bool) -> Result,
+    {
+        let mut current_position = 0usize;
+        let mut last_position = 0usize;
+        let mut max_position = 0usize;
+        let mut should_print = true;
+        loop {
+            self.visit_nested_tables(path, is_array_of_tables, &mut |t, path, is_array| {
+                match &t.position {
+                    Some(ref pos) => {
+                        if *pos > max_position {
+                            max_position = *pos;
+                        }
+                        if *pos == current_position {
+                            // if this table is from from the position we're
+                            // looking for, print it.
+                            should_print = true;
+                            current_position += 1;
+                        } else {
+                            // if this table isn't from the position we're
+                            // looking for, skip it.
+                            should_print = false;
+                        }
+                    }
+                    // If this table doesn't have a position then it was
+                    // probably made programmatically, so put it after wherever
+                    // it happens to be after in the tree. We only want to print
+                    // it once though, so rely on should_print from the previous
+                    // table to tell you whether you need to print this one.
+                    None => (),
```

Comment:
I have just thought of a case where this might do a surprising thing. If you parse this:
```
[a]
[b]
[a.z]
```
and then add a table `[a.y]` programatically, it will end up printing
```
[a]
[b]
[a.z]
[a.y]
```
when you were expecting:
```
[a]
[a.y]
[b]
[a.z]
```
I think that this is kind-of fine though: if you want to force the order then you can place `[a.y]` manually before `[a.z]` and it will go in the right place. Alternatively, we could infer should_print from the parent rather than the sibling when we find a Table with position=None and it would do the thing you were expecting. This might mean that we need to re-think our use of the visitor pattern though.

Reply:
In practice, I think that this is what you want. If you have
[package]
[dependencies]
[[[example]]
[dependencies.opencl]
[dev-dependencies]
```
and you add `[dependencies.somethingelse]` then you probably won't be surprised if it inserts itself after `[dependencies.opencl]`.